File: backend/app/services/lego_blocks/hierarchy_path_block.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path, PurePosixPath
import shutil
import sqlite3
from uuid import uuid4


def _now_iso_block() -> str:
    return datetime.now(timezone.utc).isoformat()


def normalize_relative_path_block(value: str) -> str:
    normalized = value.strip().replace("\\", "/")
    if not normalized:
        raise ValueError("Path cannot be empty")
    posix = PurePosixPath(normalized)
    if posix.is_absolute() or ".." in posix.parts:
        raise ValueError("Path must be vault-relative")
    return posix.as_posix()


def resolve_vault_relative_path_block(vault_root: Path, relative_path: str) -> Path:
    clean_rel = normalize_relative_path_block(relative_path)
    root = vault_root.resolve()
    resolved = (root / clean_rel).resolve()
    resolved.relative_to(root)
    return resolved


def ensure_node_markdown_file_block(
    vault_root: Path,
    *,
    relative_path: str,
    node_type: str,
    title: str,
) -> None:
    target = resolve_vault_relative_path_block(vault_root, relative_path)
    target.parent.mkdir(parents=True, exist_ok=True)
    if target.exists():
        return
    content = f"# {title.strip()}\n\n<!-- type: {node_type} -->\n"
    target.write_text(content, encoding="utf-8")


def copy_and_archive_path_transition_block(
    vault_root: Path,
    *,
    from_relative_path: str,
    to_relative_path: str,
) -> bool:
    from_abs = resolve_vault_relative_path_block(vault_root, from_relative_path)
    to_abs = resolve_vault_relative_path_block(vault_root, to_relative_path)
    if from_abs == to_abs:
        return False
    if not from_abs.exists():
        return False
    if to_abs.exists():
        raise FileExistsError(f"Cannot copy to existing path: {to_relative_path}")

    to_abs.parent.mkdir(parents=True, exist_ok=True)
    shutil.copy2(from_abs, to_abs)

    now = datetime.now(timezone.utc)
    archive_rel = (
        Path(".ltm-pilot")
        / "archive"
        / now.strftime("%Y-%m-%d")
        / now.strftime("%H%M%S-%f")
        / normalize_relative_path_block(from_relative_path)
    )
    archive_abs = resolve_vault_relative_path_block(vault_root, str(archive_rel))
    archive_abs.parent.mkdir(parents=True, exist_ok=True)
    shutil.move(str(from_abs), str(archive_abs))
    return True


def upsert_path_alias_block(
    conn: sqlite3.Connection,
    *,
    alias_path: str,
    target_type: str,
    target_id: str,
) -> None:
    clean_alias = normalize_relative_path_block(alias_path)
    conn.execute(
        """
INSERT INTO path_aliases (id, alias_path, target_type, target_id, is_active, created_at)
VALUES (?, ?, ?, ?, 1, ?)
ON CONFLICT(alias_path) DO UPDATE SET
  target_type = excluded.target_type,
  target_id = excluded.target_id,
  is_active = 1
""".strip(),
        (str(uuid4()), clean_alias, target_type, target_id, _now_iso_block()),
    )


@dataclass(frozen=True)
class PathResolutionBlock:
    requested_path: str
    resolved_path: str
    target_type: str
    target_id: str
    via_alias: bool

    def to_dict(self) -> dict:
        return {
            "requested_path": self.requested_path,
            "resolved_path": self.resolved_path,
            "target_type": self.target_type,
            "target_id": self.target_id,
            "via_alias": self.via_alias,
        }


def resolve_path_block(conn: sqlite3.Connection, requested_path: str) -> PathResolutionBlock | None:
    clean_path = normalize_relative_path_block(requested_path)

    node = conn.execute(
        "SELECT id, file_path FROM nodes WHERE file_path = ? LIMIT 1",
        (clean_path,),
    ).fetchone()
    if node is not None:
        return PathResolutionBlock(
            requested_path=clean_path,
            resolved_path=str(node["file_path"]),
            target_type="node",
            target_id=str(node["id"]),
            via_alias=False,
        )

    thought = conn.execute(
        "SELECT id, file_path FROM thoughts WHERE file_path = ? LIMIT 1",
        (clean_path,),
    ).fetchone()
    if thought is not None:
        return PathResolutionBlock(
            requested_path=clean_path,
            resolved_path=str(thought["file_path"]),
            target_type="thought",
            target_id=str(thought["id"]),
            via_alias=False,
        )

    alias = conn.execute(
        """
SELECT alias_path, target_type, target_id
FROM path_aliases
WHERE alias_path = ? AND is_active = 1
LIMIT 1
""".strip(),
        (clean_path,),
    ).fetchone()
    if alias is None:
        return None

    target_type = str(alias["target_type"])
    target_id = str(alias["target_id"])
    if target_type == "node":
        target = conn.execute("SELECT file_path FROM nodes WHERE id = ? LIMIT 1", (target_id,)).fetchone()
    else:
        target = conn.execute("SELECT file_path FROM thoughts WHERE id = ? LIMIT 1", (target_id,)).fetchone()
    if target is None:
        return None

    return PathResolutionBlock(
        requested_path=clean_path,
        resolved_path=str(target["file_path"]),
        target_type=target_type,
        target_id=target_id,
        via_alias=True,
    )

File: frontend/electron/src/lego_blocks/hierarchyPathBlock.ts
import * as fs from 'fs'
import * as path from 'path'

function normalizeRelativePathBlock(value: string): string {
  const normalized = value.trim().replace(/\\/g, '/')
  if (!normalized) throw new Error('Path cannot be empty')
  if (normalized.startsWith('/')) throw new Error('Path must be vault-relative')
  const parts = normalized.split('/').filter(Boolean)
  if (parts.includes('..')) throw new Error('Path must be vault-relative')
  return parts.join('/')
}

function resolveVaultRelativePathBlock(vaultRoot: string, relativePath: string): string {
  const cleanRel = normalizeRelativePathBlock(relativePath)
  const rootResolved = path.resolve(vaultRoot)
  const resolved = path.resolve(vaultRoot, cleanRel)
  if (!resolved.startsWith(rootResolved + path.sep) && resolved !== rootResolved) {
    throw new Error('Path traversal detected')
  }
  return resolved
}

export function ensureNodeMarkdownFileBlock(params: {
  vaultRoot: string
  relativePath: string
  nodeType: string
  title: string
}): void {
  const target = resolveVaultRelativePathBlock(params.vaultRoot, params.relativePath)
  fs.mkdirSync(path.dirname(target), { recursive: true })
  if (fs.existsSync(target)) return
  const content = `# ${params.title.trim()}\n\n<!-- type: ${params.nodeType} -->\n`
  fs.writeFileSync(target, content, 'utf-8')
}

export function copyAndArchivePathTransitionBlock(params: {
  vaultRoot: string
  fromRelativePath: string
  toRelativePath: string
}): boolean {
  const fromAbs = resolveVaultRelativePathBlock(params.vaultRoot, params.fromRelativePath)
  const toAbs = resolveVaultRelativePathBlock(params.vaultRoot, params.toRelativePath)
  if (fromAbs === toAbs) return false

  if (!fs.existsSync(fromAbs)) return false
  if (fs.existsSync(toAbs)) throw new Error(`Cannot copy to existing path: ${params.toRelativePath}`)

  fs.mkdirSync(path.dirname(toAbs), { recursive: true })
  fs.copyFileSync(fromAbs, toAbs)

  const now = new Date()
  const y = now.getUTCFullYear()
  const m = String(now.getUTCMonth() + 1).padStart(2, '0')
  const d = String(now.getUTCDate()).padStart(2, '0')
  const hh = String(now.getUTCHours()).padStart(2, '0')
  const mm = String(now.getUTCMinutes()).padStart(2, '0')
  const ss = String(now.getUTCSeconds()).padStart(2, '0')
  const ms = String(now.getUTCMilliseconds()).padStart(3, '0')

  const archiveRel = path.join(
    '.think-space',
    'archive',
    `${y}-${m}-${d}`,
    `${hh}${mm}${ss}-${ms}`,
    normalizeRelativePathBlock(params.fromRelativePath),
  )
  const archiveAbs = resolveVaultRelativePathBlock(params.vaultRoot, archiveRel)
  fs.mkdirSync(path.dirname(archiveAbs), { recursive: true })
  fs.renameSync(fromAbs, archiveAbs)
  return true
}

export function normalizeHierarchyRelativePathBlock(value: string): string {
  return normalizeRelativePathBlock(value)
}
